opal/class/opal_lifo.h

```diff
@@ -104,8 +104,47 @@ static inline bool opal_lifo_is_empty( opal_lifo_t* lifo )
     return (opal_list_item_t *) lifo->opal_lifo_head.data.item == &lifo->opal_lifo_ghost;
 }
 
+#if OPAL_HAVE_ATOMIC_LLSC_PTR
 
-#if OPAL_HAVE_ATOMIC_CMPSET_128
+static inline opal_list_item_t *opal_lifo_push_atomic (opal_lifo_t *lifo,
+                                                       opal_list_item_t *item)
+{
+    opal_list_item_t *next;
+
+    do {
+        item->opal_list_next = next = (opal_list_item_t *) opal_atomic_ll_ptr (&lifo->opal_lifo_head.data.item);
+        opal_atomic_wmb ();
```

Comment:
The barrier is meant to ensure the ordering between the store instruction for writing item->opal_list_next and the store conditional. I think there is an example of something like this in the PPC spec Paul pointed out.

I don't think the extra store will clear the reservation. Though I could reorder the loop to do a normal read/store to update opal_list_next. Then do a ldarx/stdcx. pair with a conditional between them that ensures the current head is the same as what was stored in the opal_list_next pointer.